cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -339,8 +375,99 @@ impl YaccParser {
+        let s = &self.src[i..];
+        for (kind_name, kind) in YACC_KINDS {
+            if starts_with_lower(kind_name, s) {
+                if update_yacc_kind {
+                    self.yacc_kind = Some(kind);
+                }
+                let end_pos = j + kind_name.len();
+                return Ok(end_pos);
+            }
+        }
+        // This strikes me as an interesting case where we know the start pos of the
+        // invalid yacc kind value, but do not know the end, we could `self.parse_to_eol()`.
+        Err(self.mk_error(YaccGrammarErrorKind::InvalidYaccKind, i))
+    }
+
+    fn parse_grmtools_header(&mut self, mut i: usize) -> Result<usize, YaccGrammarError> {
+        let mut yacc_kind_key_span = None;
+        let update_yacc_kind = !matches!(self.yacc_kind_resolver, YaccKindResolver::Force(_));
+        let require_yacc_kind = matches!(self.yacc_kind_resolver, YaccKindResolver::NoDefault);
+
+        i = self.parse_ws(i, true)?;
+        if let Some(j) = self.lookahead_is("%grmtools", i) {
+            i = self.parse_ws(j, true)?;
+            if let Some(j) = self.lookahead_is("{", i) {
+                i = self.parse_ws(j, true)?;
+                while self.lookahead_is("}", i).is_none() {
+                    let key_start_pos = i;
+                    let (key_end_pos, key) = self.parse_name(i)?;
+                    i = self.parse_ws(key_end_pos, false)?;
+                    if key == "yacckind" {
+                        let val_end_pos = self.parse_yacckind(i, update_yacc_kind)?;
+                        if let Some(orig) = yacc_kind_key_span {
+                            let dupe = Span::new(key_start_pos, key_end_pos);
+                            return Err(YaccGrammarError {
+                                kind: YaccGrammarErrorKind::DuplicateGrmtoolsHeaderKey,
+                                spans: vec![orig, dupe],
+                            });
+                        } else {
+                            yacc_kind_key_span = Some(Span::new(key_start_pos, key_end_pos));
+                            i = self.parse_ws(val_end_pos, true)?;
+                        }
+                    } else {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::InvalidGrmtoolsHeaderKey,
+                            spans: vec![Span::new(key_start_pos, key_end_pos)],
+                        });
+                    }
+                }
+                if let Some(i) = self.lookahead_is("}", i) {
+                    return Ok(i);
+                }
+            } else if require_yacc_kind {
+                return Err(self.mk_error(YaccGrammarErrorKind::MissingGrmtoolsHeader, i));
+            }
+        } else if require_yacc_kind {
+            return Err(self.mk_error(YaccGrammarErrorKind::MissingGrmtoolsHeader, i));
+        }
+        Ok(i)
+    }
+
+    pub(crate) fn finish(self) -> (Option<YaccKind>, GrammarAST) {
```

Comment:
I wonder if we should call this `build` -- it's a more common name for this sort of Rust thing? But perhaps `build` isn't quite right, and `finish` (or maybe `to_ast`?) is better?

Reply:
`build` seems good to me, in the why didn't I think of that way.

Reply:
should be fixed in [c84c3f2](https://github.com/softdevteam/grmtools/pull/492/commits/c84c3f28530f09b49df6b9dff8089b3ef35b368f)